WLC/code_fixing/codefixer.py

```diff
@@ -422,46 +521,95 @@ def fix_func_or_class_call(self, match, poss_chars):
 
     def fix_method_call(self, match, poss_chars):
         groups = match.groups()
-        poss_var = poss_chars[match.start(2): match.end(2)]
-        closest_var, _ = self.levenshtein_closest(poss_var, self.context["variables"])
-        LOGGER.debug("Fixing method call var. From {} to {}".format(groups[1], closest_var))
+        poss_callable = poss_chars[match.start(2): match.end(2)]
+        poss_method = poss_chars[match.start(3): match.end(3)]
+
+        if self.curr_def:
+            ctxt = self.def_context[self.curr_def]['variables'] + self.context['variables']
+        elif self.curr_class:
+            ctxt = self.class_context[self.curr_class]['variables'] + self.context['variables']
+        else:
+            ctxt = self.context['variables']
+
+        closest_var, distance_var = self.levenshtein_closest(poss_callable, ctxt)
+        closest_import, distance_import = self.levenshtein_closest(poss_callable, self.context["imports"])
+
+        # Check if its more likely to be a call on a custom variable, or an import.
+        if distance_var <= distance_import:
+            # is a var -> harder to check method -> TODO?
+            closest_callable = closest_var
+            LOGGER.debug("Fixing method call var. From {} to {}".format(groups[1], closest_callable))
+
+            closest_method, _ = self.levenshtein_closest(poss_method, self.context["methods"])
+            LOGGER.debug("Fixing method call method. From {} to {}".format(groups[2], closest_method))
+        else:
+            # is an import -> method can be checked if __all__ is present
+            # TODO: workaround __all__ not being present
+            closest_callable = closest_import
+            LOGGER.debug("Method call was found to be on an import!")
+            LOGGER.debug("Fixing method call var. From {} to {}".format(groups[1], closest_import))
+
+            imported = importlib.import_module(closest_import)
+            if hasattr(imported, "__all__"):
+                closest_method, _ = self.levenshtein_closest(poss_method, imported.__all__)
+            else:
+                closest_method = groups[2]
+            del imported
+
+            LOGGER.debug("Fixing method call method. From {} to {}".format(groups[2], closest_method))
+
+        closest_args = self.fix_arguments(groups[3], poss_chars[match.start(4): match.end(4)])
+
+        LOGGER.debug("Fixing method call args. From {} to {}.".format(groups[3], closest_args))
+
+        return '{}.{}({})'.format(closest_callable, closest_method, closest_args)
+
+    def fix_self_method_call(self, match, poss_chars):
+        # NOTE: 'self' is group 1, so as to be able to pass it to standard method fixing
+        groups = match.groups()
+        # Not in class -> shouldn't be self, revert to fixing
```

Comment:
Is this a good idea? If the person puts self at the start should we just let the person do it and then fail?

Reply:
What do you mean? Do you still think we should fix it to self outside of a class? if they really did mean to type self, itll fix it to that anyway, so i think this is fine.

Reply:
Well currently if they didnt mean to say self and they are not in a class then it will go to the normal fixing which will remove self no? But I though if they meant to use self then it should be kept even if they made a mistake.

Reply:
It won't remove it, it'll just try to fix it to something other than self - if that doesn't work then it moves on happily. It's not overdoing the fixing like Mark was worried about, if that's what you're thinking